### ISSUE TITLE
systemd oneshots

### ISSUE THREAD

**Issue by bcwaldon** [coreos/fleet#562]
We need real support for systemd oneshot units. Not sure if this is different from the concept of a "batch job".

**Comment by OAGr** [comment]
Any update on this?  I'm also using oneshot services for a continuous integration system. Right now they ping back a remote server on completion, which sends a message to fleet to destroy the service.

**Comment by bcwaldon** [comment]
@OAGr tl;dr: no

Support for oneshots can't easily be layered in, as fleet's operation is built around a declarative model. For example, "I want foo.service to be launched somewhere". I'm more than happy to work with someone to figure out how we can properly support oneshots, but it will be nontrivial.

**Comment by ngauthier** [comment]
Just to give more feedback and info here, we're using oneshots also as part of CI. We are also doing it via the HTTP API.

To find out if a oneshot has started, we can look at:

```go
unit, _ := api.Unit(service)
unit.CurrentState
```

However, there's no distinction between being started and finishing. For that, we use the direct systemd state call:

```go
service := "myoneshot.service"

unitStates, _ := api.UnitStates()

var unitState *schema.UnitState
for _, s := range unitStates {
	if s.Name == service {
		unitState = s
		break
	}
}

if unitState == nil {
	// we don't know about this unit
	return errors.New("unit not in state list")
}

if unitState.SystemdActiveState == "inactive" {
	// unit has stopped well
} else {
	// unit hasn't stopped
}
```

This is far less optimal because we have to ask for every service's state then find our service and check its systemd state.

Would be awesome to extend Unit in the schema with a systemd state attribute so the `api.Unit(service)` call would return systemd state.

Or, as mentioned here, have a fleet state for a completed oneshot service or something similar.

**Comment by bcwaldon** [comment]
@ngauthier I don't believe the suggestion to provide unit state in the Unit() call is relevant to the issue of native oneshot support... unless I'm missing something. 

As I stated above, fleet is sort of designed against oneshots right now. Do you have an idea of how oneshots could be better supported through the fleet internals?

**Comment by ngauthier** [comment]
I kind of considered it part of native oneshot support because the fleet status doesn't reflect the actual oneshot status (systemd's representation of oneshot status is richer than fleets, so it can represent the completed oneshot, which is why I use systemd's state). The main reason I posted the example was because I ended up on this thread trying to work with oneshots in fleet, so I hoped it would help others.

As far as fleet's internals, I am probably too inexperienced to suggest an amazing solution, but one thought I had was to show a completed oneshot as being in the loaded state when it's done, not in the activate state. So it ran and it's done and now it's just loaded. But I guess it would be hard to tell if it ran at all.

**Comment by xied75** [comment]
Dear all I would like to verify with you guys about a use case:

I've got a cluster up and running, and my fleet unit need to git clone a private repo, which means I need to get a private key for root on each of the node so that they can access my gitlab via Deploykey.

In an ideal world, you do this in the cloud-config.yaml, assume that you knows everything needed before you deploy the cluster. But you always find something is missing from the cluster later.

I was also thinking should this be a job rather for Ansible/Puppet, in that I want to 'state' control my nodes to be in a specific 'state'.

But I tend to reduce the tooling needed to have a unified story, i.e. I'll write a simple fleet unit to dump the private key for root. Which apparently is 'oneshot', and global. And it should disappear afterwards.

Even further, what if we could do 
```
fleetctl run whatever-arbitrary-bash
```
Then fleetd will magically turn this into a oneshot unit and fire it.

**Comment by djmaze** [comment]
Just as a side note because this seems not known widely: You can make the server re-evaluate a cloud config file without rebooting using the `coreos-cloudinit` binary. This is particularly useful if you have a config drive which can be modified on-the-fly (e.g. with qemu on the host):

    sudo coreos-cloudinit --from-configdrive=/media/configvirtfs

So you just need a way to fire that command all servers. Could be made a global oneshot unit..

I admit this does not help with most cloud providers though, as they don't let you modify the cloud config of a running server.

**Comment by rdark** [comment]
@xied75  - I solved that problem (I actually reached it first by hitting maximum user-data size on AWS, but it solves this issue also) by chaining additional cloudinit scripts from an external URL (in this case it's stored in an s3 bucket reached via a VPC endpoint. This is what's in my user-data populated cloudinit:

    - name: cloudinit-includes.service
      command: start
      content: |
        [Unit]
        Description=Install External Cloudinit Scripts
        ConditionPathExists=/etc/custom_environment
        [Service]
        EnvironmentFile=/etc/custom_environment
        Type=oneshot
        RemainAfterExit=yes
        ExecStart=/usr/bin/coreos-cloudinit -from-url ${CLOUDINIT_INCL_URL}/coreos/common/cloud_config-coreos_units_cloudinit-includes.yaml

I have a bucket for each cluster, content of which is stored in git, with a branch for each cluster that corresponds to a bucket. The bucket gets populated via a git post-recieve hook, so I can roll out changes in a semi-controlled fashion. The above URL gets polled at every boot (or restart of cloudinit), and pulls down a chainloader for other cloud-init things.

    #cloud-config
    coreos:
      units:
        - name: install-cloudinit-includes.service
          command: start
          content: |
            [Unit]
            Description=Cloudinit External Units
            ConditionPathExists=/etc/custom_environment
            [Service]
            Type=oneshot
            EnvironmentFile=/etc/custom_environment
            RemainAfterExit=yes
            ExecStart=/usr/bin/bash -c '/usr/bin/coreos-cloudinit -from-url ${CLOUDINIT_INCL_URL}/coreos/common/install-ca-cert.pem.yaml && \
                                        /usr/bin/coreos-cloudinit -from-url ${CLOUDINIT_INCL_URL}/coreos/common/write_files-iptables-rules-save.yaml && \
                                        /usr/bin/coreos-cloudinit -from-url ${CLOUDINIT_INCL_URL}/coreos/common/cloud_config-coreos_units_update-custom-ca-certificates.yaml && \
                                        /usr/bin/coreos-cloudinit -from-url ${CLOUDINIT_INCL_URL}/coreos/common/cloud_config-coreos_units_install_etcd_backup.yaml && \
                                        /usr/bin/coreos-cloudinit -from-url ${CLOUDINIT_INCL_URL}/coreos/common/cloud_config-coreos_units_install_jq_lookup.yaml && \
                                        /usr/bin/coreos-cloudinit -from-url ${CLOUDINIT_INCL_URL}/coreos/common/cloud_config-coreos_units_install_etcdctl_hosts_lookup.yaml && \
                                        /usr/bin/coreos-cloudinit -from-url ${CLOUDINIT_INCL_URL}/coreos/common/cloud_config-coreos_units_install_etcdctl_get_coreos_hosts.yaml && \
                                        /usr/bin/coreos-cloudinit -from-url ${CLOUDINIT_INCL_URL}/coreos/common/cloud_config-coreos_units_iptables-restore-custom.yaml'

**Comment by lynchc** [comment]
+1 @xied75 

the idea of having 
```fleetctl run whatever-arbitrary-bash``` would solve all of this. I use a lot of oneshots for CI and deploying. Having it take some filtering on machine metadata as advanced option would really be nice. I can't imagine this would be too difficult and sounds like it would solve all of these problems, no?

**Comment by jgunthorpe** [comment]
I also would like to see one shot units work better.

@bcwaldon - All declarative models work by forcing a current state to a desired state, so it appears the missing element for one short support is a persistent log of one shot completion timestamp (ie the state component, one shots cannot use the 'current system' as a state).

A reasonable model is for fleet to monitor the one shot service after starting it. When it sees it has completed, it records the completion timestamp in a persistent global log (etcd I guess?). The declarative forcing function for one shots is then "start if unit X never completed after requested start time Y, else stop". real time is used to provide a fence for allowing the user to trigger the one shot repeatedly.

That is all pretty simple for singleton units (the 'log' is just a single value), global units  are a bit more complex and require a storing a map of machine id and completion time ..
